rs/execution_environment/src/scheduler.rs

```diff
@@ -2338,40 +2333,89 @@ fn join_consumed_cycles_by_use_case(
     }
 }
 
-/// Helper function that checks if a message can be executed:
+/// Helper function that checks if a subnet message can be executed:
 ///     1. A message cannot be executed if it is directed to a canister
 ///     with another long-running execution in progress.
 ///     2. Install code messages can only be executed sequentially.
-fn can_execute_msg(
+fn can_execute_subnet_msg(
     msg: &CanisterMessage,
     ongoing_long_install_code: bool,
-    long_running_canister_ids: &BTreeSet<CanisterId>,
+    canister_states: &BTreeMap<CanisterId, CanisterState>,
 ) -> bool {
-    if let Some(effective_canister_id) = msg.effective_canister_id() {
-        if long_running_canister_ids.contains(&effective_canister_id) {
-            return false;
-        }
+    let Some(effective_canister_id) = msg.effective_canister_id() else {
+        // If there is no effective canister ID, we can execute the subnet message.
+        return true;
+    };
+    let Some(effective_canister_state) = canister_states.get(&effective_canister_id) else {
+        // If there is no effective canister state, we can execute the subnet message.
+        return true;
+    };
+    if effective_canister_state.has_paused_execution() {
+        // If there is a paused execution, we can't execute the subnet message.
+        // Note, it does NOT include aborted executions.
+        return false;
     }
+    let maybe_method = match msg {
+        CanisterMessage::Ingress(ingress) => {
+            Ic00Method::from_str(ingress.method_name.as_str()).ok()
+        }
+        CanisterMessage::Request(request) => {
+            Ic00Method::from_str(request.method_name.as_str()).ok()
+        }
+        CanisterMessage::Response(_) => None,
+    };
+    let Some(method) = maybe_method else {
+        // If there is no method name, we can execute the subnet message.
+        return true;
+    };
 
-    if ongoing_long_install_code {
-        let maybe_install_code_method = match msg {
-            CanisterMessage::Ingress(ingress) => {
-                Ic00Method::from_str(ingress.method_name.as_str()).ok()
-            }
-            CanisterMessage::Request(request) => {
-                Ic00Method::from_str(request.method_name.as_str()).ok()
-            }
-            CanisterMessage::Response(_) => None,
-        };
-
+    let effective_canister_paused_or_aborted = match effective_canister_state.next_execution() {
+        NextExecution::None | NextExecution::StartNew => false,
+        NextExecution::ContinueLong | NextExecution::ContinueInstallCode => true,
+    };
+    match method {
+        // It's safe to allow other subnet on aborted canisters messages,
+        // but for the sake of minimizing the change, allow just this for now...
+        Ic00Method::DepositCycles => true,
```

Comment:
I think if it's safe to execute messages on aborted canisters, then let's do the rest that we can. We are making the change anyway, switching the match statement later doesn't buy us much, does it?